src/main/scala/routemap.scala

```diff
@@ -34,39 +35,125 @@ object RouteMap {
+      return OpenFlights.routes(airportSrc.code)(airportDest.code)
+        .map(r => GeoRoute(
+          r,
+          airportSrc.lat, airportSrc.lng,
+          airportDest.lat, airportDest.lng
+        ))
+    else
+      return List[GeoRoute]()
+  }
 
   /**
    * Find all indirect routes between two airports
    * There is at least one connecting airport between the two.
    */
-  def findIndirectAirportRoutes(airportSrc: String, airportDest: String, maxDegree: Int): List[Route] = {
+  def findIndirectAirportRoutes(airportSrc: Airport, airportDest: Airport, maxDegree: Int): List[List[GeoRoute]] = {
 
-    // TAOTODO:
-    return List[Route]()
+    // Perform a **greedy** depth-first-search
+    return findConnectingRoutes(
+      airportSrc,
+      airportDest,
+      maxDegree,
+      List()
+    )
+  }
+
+  /**
+   * Calculate a total distance in metre of the routes
+   */
+  private def totalRouteDistance(routes: List[GeoRoute]): Float = {
+    routes.map(_.distance).sum
+  }
+
+  /**
+   * Find routes which connect two airports.
+   * The function may return multiple possible routes which it finds.
+   */
+  private def findConnectingRoutes(airportSrc: Airport, airportDest: Airport, maxDegree: Int, prevRoute: List[GeoRoute]): List[List[GeoRoute]] = {
+    // Stopping criterion
+    if (maxDegree <= 0)
+      return List()
+
+    // TAOTODO: Other stopping criteria:
+    // - Too large total distance
+    // - Too large one single distance of a flight
+
+    println(Console.CYAN + "==================================" + Console.RESET)
+    println(Console.CYAN + s"[Max ${maxDegree} degrees left]..." + Console.RESET)
+    println(Console.GREEN + "Expanding route from: " + prevRoute.map(_.route.airportSourceCode).mkString(" ➡️ ") +
+      " " + airportSrc.code + Console.RESET)
+    println(Console.CYAN + "==================================" + Console.RESET)
+
+    // Expand all routes which start at the source airport
+    // (Depth-first)
+    val routesFromSrc = OpenFlights.routes(airportSrc.code)
+
+    routesFromSrc.foldLeft(List[List[GeoRoute]]()) { (allRoutes, n) =>
+
+      val (nextDest, routes) = n
+      var allRoutes_ = List[List[GeoRoute]]()
+
+      // Skip backward routes or invalid routes
+      if (OpenFlights.airports.contains(nextDest) && nextDest != airportSrc.code) {
+        val nextDestAirport = OpenFlights.airports(nextDest)
+
+        println(airportSrc.code + " to " + nextDest + s" ${routes.length} routes")
+
+        // These geolocations are useful for future implementation
+        val (lat0, lng0) = (airportSrc.lat, airportSrc.lng)
+        val (lat1, lng1) = (nextDestAirport.lat, nextDestAirport.lng)
+
+        // Extend the previous routes with the expanded route list
+
+        // ENHANCEMENT: List is not good at appending
+        // Needs to find a better list-like structure
+        val nextRoutes = routes.map(r =>
+          prevRoute ++ List(GeoRoute(r, lat0, lng0, lat1, lng1)))
```

Comment:
If you have one list that you append to repeatedly, the go-to for that case is usually `ArrayBuffer`. It is typically idiomatic to contain use of mutable structures within single methods as much as possible—use it and then call `toList` on it when you're finished, and return that. In that case, you might use `ListBuffer` instead, which also has constant-time append. See [this set of docs](http://docs.scala-lang.org/overviews/collections/concrete-mutable-collection-classes.html) for help making those decisions. 

Also mind that you're actually doing a concat of two lists here rather than appending a single-item, which in a loop is likely much worse—you probably want something like `prevRoute :+ GeoRoute(r, lat0, lng0, lat1, lng1))` here for the time being. Appending is probably not a big deal in the case of routes anyway, where the list is never going to be very large—it might very well be more expensive to create a mutable structure and then convert it back to a `List` than to prepend on a `List` and then `reverse` it when finished, or just append to a small `List`.

General Scala mindset: do everything immutably until you really find that it's necessary to change. The immutable data structures are often more efficient than you think.